### PR TITLE
Feature/0296 possibility to update video streaming capabilities

### LINKED ISSUE #290: [SDL 0296] Possibility to update video streaming capabilities during ignition cycle
Proposal: [Possibility to update video streaming capabilities during ignition cycle](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0296-Update-video-streaming-capabilities-during-ignition-cycle.md)

> The main purpose of this proposal is to make it possible to update video streaming capabilities (VSC) during the ignition cycle.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/983

**Steering Committee Decision:**

The Steering Committee voted to accept this proposal with the following revisions:

- (Item 3) For iOS, change `streamingVideoDelegate` to `delegate`.  For Both iOS and Java Suite, use `VideoStreamingRange`.
- (Item 10) Use `VideoStreamingCapability` struct for the AppCapability feature.  While each capability will have to be reviewed to see if they can be reused for both system and app capabilities, we do feel confident enough to reuse `VideoStreamingCapability` for both.
- (Item 11) Add the following to the HMI Integration Guidelines, and extend the `Resolution Switching` diagram to document this behavior.
> The HMI should stop displaying video to the user after resizing the window instead of waiting for the StopStream request from Core. Video would resume after the video service is restarted. See the following example state flow: "user attempts to resize" -> "HMI stops displaying video" -> "head unit sends OnAppCapabilityUpdated" -> "HMI resizes/waits for stop and start of video stream" -> "display new video stream".
- (Item 12) Add `HMI` as an impacted platform.

The proposal .md file was updated to reflect these revisions on 7/27/2020.

### PR DESCRIPTION
**Pull request by vladmu**
Fixes #290

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior.

#### Unit Tests
Added unit tests cover [SDL-0296](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0296-Update-video-streaming-capabilities-during-ignition-cycle.md) changes

#### Core Tests
* each of the properties in the struct of `AppCapability` can be filled and send in `OnAppCapabilityUpdated` notification
* the `OnAppCapabilityUpdated` notification successfully sent
* the property `additionalVideoStreamingCapabilities` can be received from the HMI

Core version / branch / commit hash / module tested against: https://github.com/smartdevicelink/sdl_core/pull/3468
HMI name / version / branch / commit hash / module tested against: https://github.com/smartdevicelink/sdl_hmi/pull/391

### Summary
Added RPC changes requested in [SDL-0296](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0296-Update-video-streaming-capabilities-during-ignition-cycle.md).

##### Breaking Changes
* `N/A`

##### Enhancements
* Added a new enum `AppCapabilityType`
* Added a new struct `AppCapability`
* Added a new message `OnAppCapabilityUpdated`
* Added getter/setter of a new property `additionalVideoStreamingCapabilities` in `VideoStreamingCapability`

##### Bug Fixes
* `N/A`

### Tasks Remaining:
- `N/A`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by vladmu** [comment]
@kshala-ford @mrapitis @atiwari9 @santhanamk please review the changes.

**Comment by santhanamk** [comment]
@vladmu Sure I will start the review once I have finished reviewing the changes in the sdl-java-suite PR.

File: tests/node/rpc/enums/AppCapabilityTypeTests.js
Comment (santhanamk) [inline tests/node/rpc/enums/AppCapabilityTypeTests.js:11]: @vladmu Do you need to use `Validator.assertNotNullUndefined(...)` here on line 11 instead of `Validator.assertNotNull(...)`?  `AmbientLightStatusTests.js` is using that in its `testValidEnums` function.

File: lib/js/src/rpc/enums/FunctionID.js
Comment (vladmu) [reply lib/js/src/rpc/enums/FunctionID.js:876]: @santhanamk I've checked this file https://github.com/LuxoftSDL/rpc_spec/blob/feature/0296_possibility_to_update_video_streaming_capabilities/MOBILE_API.xml and noticed `OnSubtleAlertPressed` is not a part of it. But the removal of `OnSubtleAlertPressed` is out of scope of the proposal. So I'm confused about how to deal with this ID conflict. `'OnAppCapabilityUpdated': 0x8014,` is correct value according to related `rpc_spec`

File: tests/node/rpc/enums/AppCapabilityTypeTests.js
Comment (vladmu) [inline tests/node/rpc/enums/AppCapabilityTypeTests.js:11]:
```suggestion
        Validator.assertNotNullUndefined(enumVideoStreaming, 'VIDEO_STREAMING returned null');
```

Comment (vladmu) [reply tests/node/rpc/enums/AppCapabilityTypeTests.js]: Done in https://github.com/smartdevicelink/sdl_javascript_suite/pull/302/commits/4b6e18664545fe2547f9019350240bf1b713ce5b

File: lib/js/src/rpc/enums/FunctionID.js
Comment (santhanamk) [reply lib/js/src/rpc/enums/FunctionID.js:876]: Hi @vladmu maybe this is a question we can ask Livio when they do the code review.

**Comment by santhanamk** [comment]
Hi @theresalech this PR is ready for Livio review.  

There is one question that we have from the code review.  If Livio can address this item during their code review that would be great.  Thanks.

File: lib/js/src/rpc/enums/FunctionID.js
Comment (santhanamk) [reply lib/js/src/rpc/enums/FunctionID.js:876]: Hi @theresalech and @crokita .  This question came up during the code review.  If you could address this during the code review that would be great.  Thanks.

**Comment by theresalech** [comment]
Hi @santhanamk, thank you for letting us know! This feature is not currently planned for the next JavaScript Suite release (1.2), so we will review once the Steering Committee has prioritized the feature for a following release.

**Review by santhanamk — APPROVED** [review]
Hi @vladmu the latest changes look good.  The issue we discovered earlier would need to be discussed when Livio does their review.

**Comment by vladmu** [comment]
Hi @JackLivio,
Because the 1.2 version was released and this feature included in the next release's scope, could you proceed with the Livio review? Also, we appreciate your help with the last open question in https://github.com/smartdevicelink/sdl_javascript_suite/pull/302#discussion_r474098165

**Comment by theresalech** [comment]
Hi @vladmu, please make sure you follow the [PR Template](https://github.com/smartdevicelink/sdl_javascript_suite/blob/develop/.github/PULL_REQUEST_TEMPLATE.md) for each repository when requesting reviews. For the JavaScript Suite repository, @crokita should be tagged when a PR is ready for Livio review. Thank you!

**Comment by vladmu** [comment]
> Hi @vladmu, please make sure you follow the [PR Template](https://github.com/smartdevicelink/sdl_javascript_suite/blob/develop/.github/PULL_REQUEST_TEMPLATE.md) for each repository when requesting reviews. For the JavaScript Suite repository, @crokita should be tagged when a PR is ready for Livio review. Thank you!

Thank you, my comment is not a request but a reminder. The PR review was requested before on 20 Aug by @santhanamk. However, you're right that Chris was not tagged there. @crokita the PR is ready for Livio review.

**Comment by theresalech** [comment]
Thanks, @vladmu! The Steering Committee will be finalizing content for the next set of releases (planned for April 14, 2021) in early December. Once content is finalized, we'll begin reviewing code contributions based on the approved 2021 Roadmap timeline.

**Comment by jordynmackool** [comment]
Hi @vladmu and @santhanamk, since the core PR for this feature is not complete, it seems that this PR would have not been tested fully. To ensure the review process goes smoothly:

1. Please avoid removing sections from the PR template as all the information requested is needed.
2. Please provide the Core branch/PR that this was tested against.

Once this PR is fully tested, tag me and Livio will review. Thank you!

**Comment by vladmu** [comment]
@jordynmackool @crokita the PR was updated with the latest `develop` and regenerated with related `rpc_spec` values, the code of the unit tests was adjusted to make it possible to pass all tests. No any critical fixes here, only some jsdocs and import paths. The description was fixed to include all sections from the template and the code was tested against the Core and HMI (corresponded links also included in the description). Please review.

File: lib/js/src/rpc/enums/FunctionID.js
Comment (vladmu) [reply lib/js/src/rpc/enums/FunctionID.js]: @santhanamk I resolve this conversation, the question is not actual anymore. With the latest update of the `rpc_spec` the function id now is `0x8017`.

**Comment by vladmu** [comment]
@crokita, the PR was updated with the latest `develop`, please don't merge this PR after the approval before this related `rpc_spec` PR https://github.com/smartdevicelink/rpc_spec/pull/272 is not merged.

**Review by crokita — COMMENTED** [review]
All the additions look good. Will wait on the equivalent core PR's approval for final testing. 

I noticed that the Java Suite's VideoStreamingParameters and JS's _VideoStreamingParameters classes are misaligned, and that the equivalent Java PR for this feature includes additional changes to that class. While not necessary to include those changes for JS's _VideoStreamingParameters class, it's something that will need to be addressed later.

**Comment by vladmu** [comment]
@crokita sounds good. Will you switch to review another PR until waiting for core here, we could again resolve merge conflicts before your review?

**Comment by crokita** [comment]
Hello @vladmu. I believe this one will be handled next: https://github.com/smartdevicelink/sdl_javascript_suite/pull/354

**Comment by crokita** [comment]
@vladmu Hello. We're finished testing using this PR. In preparation for potentially merging this PR today could you resolve the remaining merge conflicts? Thank you

**Comment by BogMW** [comment]
Hello @crokita. All conflicts were resolved.

**Review by crokita — APPROVED** [review]

**Pull request merged**
